File: src/hive_bot/engine/state.py
"""Plain-data board state for the fast Hive engine.

Deliberately dict/tuple/int based (no pydantic) -- see constants.py and the
plan doc for why. `apply.py` mutates a `GameState` in place (make/unmake)
rather than copying it, so this module also owns the small set of read-only
helpers (`neighbors`, `occupied`, `top_piece_at`, ...) that both move
generation and move application share.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from .constants import BASE_PIECE_TYPES, HEX_DIRS, PIECE_COUNTS, PieceType

Pos = tuple[int, int, int]
PositionKey = tuple[
    tuple[tuple[Pos, tuple[int, ...]], ...], tuple[int, ...], tuple[int, ...], int
]

DRAW = -1  # sentinel for GameState.winner on a draw


def neighbors(pos: Pos) -> tuple[Pos, ...]:
    q, r, s = pos
    return tuple((q + dq, r + dr, s + ds) for dq, dr, ds in HEX_DIRS)


def shared_neighbors(a: Pos, b: Pos) -> tuple[Pos, Pos]:
    """The two hexes adjacent to both `a` and `b` (always exactly 2 for
    adjacent `a`/`b` on a hex grid) -- the pair a slide must not be
    wedged between."""
    na = set(neighbors(a))
    nb = set(neighbors(b))
    shared = na & nb
    it = iter(shared)
    return next(it), next(it)


def position_key(state: GameState) -> PositionKey:
    """A hashable snapshot of "the position" for repetition detection:
    board contents (including full stacks, since a beetle's height matters)
    plus each hand's contents plus whose turn it is. Doesn't account for the
    pillbug freeze state (which piece moved last) -- a pragmatic
    simplification, since this only needs to bound self-play game length,
    not be tournament-rules-perfect."""
    board_items = tuple(sorted((pos, tuple(stack)) for pos, stack in state.board.items()))
    hands = (tuple(sorted(state.hand[0])), tuple(sorted(state.hand[1])))
    return (board_items, *hands, state.current_player)


def positions_connected(occupied: frozenset[Pos]) -> bool:
    """Whether every position in `occupied` is reachable from every other
    via adjacency -- the "hive must stay one piece" rule. Pure graph
    connectivity over positions; piece identity doesn't matter."""
    if not occupied:
        return True
    start = next(iter(occupied))
    seen = {start}
    stack = [start]
    while stack:
        cur = stack.pop()
        for nb in neighbors(cur):
            if nb in occupied and nb not in seen:
                seen.add(nb)
                stack.append(nb)
    return len(seen) == len(occupied)


@dataclass(frozen=True, slots=True)
class Piece:
    id: int
    piece_type: PieceType
    owner: int  # 0 or 1


def build_pieces(enabled_types: frozenset[PieceType]) -> list[Piece]:
    pieces = []
    pid = 0
    for owner in (0, 1):
        for piece_type in PieceType:
            if piece_type not in enabled_types:
                continue
            for _ in range(PIECE_COUNTS[piece_type]):
                pieces.append(Piece(id=pid, piece_type=piece_type, owner=owner))
                pid += 1
    return pieces


@dataclass(slots=True)
class GameState:
    pieces: dict[int, Piece]
    board: dict[Pos, list[int]]  # occupied cells only, bottom -> top piece ids
    position: dict[int, Pos | None]  # None means still in hand
    hand: list[list[int]]  # hand[owner] = piece ids not yet placed
    queen_placed: list[bool]  # queen_placed[owner]
    current_player: int
    turn_no: int
    ply: int
    last_moved_piece_id: int | None = None
    last_moved_ply: int | None = None
    game_over: bool = False
    winner: int | None = None  # 0, 1, DRAW, or None while ongoing

    # How many times each distinct position (board + hands + side to move)
    # has occurred, for threefold-repetition draw detection. Not an official
    # Hive rule -- see constants.py's REPETITION_LIMIT.
    position_counts: dict[PositionKey, int] = field(default_factory=dict)

    @classmethod
    def new_game(cls, enabled_types: frozenset[PieceType] = BASE_PIECE_TYPES) -> GameState:
        pieces_list = build_pieces(enabled_types)
        pieces = {p.id: p for p in pieces_list}
        hand: list[list[int]] = [[], []]
        for p in pieces_list:
            hand[p.owner].append(p.id)
        return cls(
            pieces=pieces,
            board={},
            position=dict.fromkeys(pieces),
            hand=hand,
            queen_placed=[False, False],
            current_player=0,
            turn_no=1,
            ply=0,
        )

    def opponent(self) -> int:
        return 1 - self.current_player

    def occupied(self) -> frozenset[Pos]:
        return frozenset(self.board.keys())

    def stack_height_at(self, pos: Pos) -> int:
        return len(self.board.get(pos, ()))

    def top_piece_at(self, pos: Pos) -> int | None:
        stack = self.board.get(pos)
        return stack[-1] if stack else None

    def stack_index_of(self, piece_id: int) -> int:
        """0-indexed height of `piece_id` within its stack (0 = ground)."""
        pos = self.position[piece_id]
        assert pos is not None
        return self.board[pos].index(piece_id)

    def board_positions(self, owner: int) -> set[Pos]:
        positions: set[Pos] = set()
        for pid in self.pieces_on_board(owner):
            pos = self.position[pid]
            assert pos is not None
            positions.add(pos)
        return positions

    def pieces_on_board(self, owner: int) -> list[int]:
        return [
            pid
            for pid, p in self.pieces.items()
            if p.owner == owner and self.position[pid] is not None
        ]

    def is_frozen(self, piece_id: int) -> bool:
        """Pillbug freeze rule: a piece that moved "last turn" can't be
        moved or thrown this ply. The `ply - 1` comparison (rather than the
        arguably more obvious `ply`) is copied verbatim from the reference
        engine's `pillbug_throw_valid` -- see tests/reference/helpers.py and
        its test_pillbug_throw_freeze_rule -- to stay byte-for-byte
        compatible with the oracle used in test_engine_vs_reference.py."""
        return self.last_moved_piece_id == piece_id and self.last_moved_ply == self.ply - 1

File: js/src/engine/state.ts
// Plain-data board state for the in-browser Hive engine -- a direct port
// of the Python package's engine/state.py. See that file's docstring for
// why the design is dict/array based rather than richer objects: `apply.ts`
// mutates a GameState in place (make/unmake) rather than cloning it.
//
// Positions are `[q, r, s]` tuples, but JS can't use array/tuple values as
// Map keys by value (only by reference) -- so everywhere Python uses a
// `Pos` as a dict/set key, this port uses `posKey(pos)` (a `"q,r,s"`
// string, matching the same convention the Django app's own JSON
// serialization and the React frontend's `types.ts` already use) instead.

import {
  BASE_PIECE_TYPES,
  ALL_PIECE_TYPES,
  HEX_DIRS,
  PIECE_COUNTS,
  PieceType,
} from "./constants.js";

export type Pos = readonly [number, number, number];
export type Owner = 0 | 1;

export const DRAW = -1; // sentinel for GameState.winner on a draw

export function posKey(pos: Pos): string {
  return `${pos[0]},${pos[1]},${pos[2]}`;
}

export function parsePosKey(key: string): Pos {
  const parts = key.split(",").map(Number);
  return [parts[0], parts[1], parts[2]];
}

export function neighbors(pos: Pos): Pos[] {
  const [q, r, s] = pos;
  return HEX_DIRS.map(([dq, dr, ds]) => [q + dq, r + dr, s + ds] as Pos);
}

/** The two hexes adjacent to both `a` and `b` (always exactly 2 for
 * adjacent `a`/`b` on a hex grid) -- the pair a slide must not be wedged
 * between. */
export function sharedNeighbors(a: Pos, b: Pos): [Pos, Pos] {
  const na = neighbors(a);
  const bKeys = new Set(neighbors(b).map(posKey));
  const shared = na.filter((p) => bKeys.has(posKey(p)));
  return [shared[0], shared[1]];
}

/** Whether every position in `occupied` is reachable from every other via
 * adjacency -- the "hive must stay one piece" rule. Pure graph
 * connectivity over positions; piece identity doesn't matter. */
export function positionsConnected(occupied: ReadonlySet<string>): boolean {
  if (occupied.size === 0) return true;
  const start = occupied.values().next().value as string;
  const seen = new Set<string>([start]);
  const stack: Pos[] = [parsePosKey(start)];
  while (stack.length > 0) {
    const cur = stack.pop() as Pos;
    for (const nb of neighbors(cur)) {
      const key = posKey(nb);
      if (occupied.has(key) && !seen.has(key)) {
        seen.add(key);
        stack.push(nb);
      }
    }
  }
  return seen.size === occupied.size;
}

export interface Piece {
  readonly id: number;
  readonly pieceType: PieceType;
  readonly owner: Owner;
}

export function buildPieces(enabledTypes: ReadonlySet<PieceType>): Piece[] {
  const pieces: Piece[] = [];
  let pid = 0;
  for (const owner of [0, 1] as const) {
    for (const pieceType of ALL_PIECE_TYPES) {
      if (!enabledTypes.has(pieceType)) continue;
      for (let i = 0; i < PIECE_COUNTS[pieceType]; i++) {
        pieces.push({ id: pid, pieceType, owner });
        pid++;
      }
    }
  }
  return pieces;
}

export class GameState {
  pieces: Map<number, Piece>;
  board: Map<string, number[]>; // posKey -> stack of piece ids, bottom -> top
  position: Map<number, Pos | null>; // null means still in hand
  hand: [number[], number[]]; // hand[owner] = piece ids not yet placed
  queenPlaced: [boolean, boolean];
  currentPlayer: Owner;
  turnNo: number;
  ply: number;
  lastMovedPieceId: number | null = null;
  lastMovedPly: number | null = null;
  gameOver = false;
  winner: number | null = null; // 0, 1, DRAW, or null while ongoing

  // How many times each distinct position (board + hands + side to move)
  // has occurred, for threefold-repetition draw detection. Not an official
  // Hive rule -- see constants.ts's REPETITION_LIMIT.
  positionCounts: Map<string, number> = new Map();

  constructor(
    pieces: Map<number, Piece>,
    board: Map<string, number[]>,
    position: Map<number, Pos | null>,
    hand: [number[], number[]],
    queenPlaced: [boolean, boolean],
    currentPlayer: Owner,
    turnNo: number,
    ply: number,
  ) {
    this.pieces = pieces;
    this.board = board;
    this.position = position;
    this.hand = hand;
    this.queenPlaced = queenPlaced;
    this.currentPlayer = currentPlayer;
    this.turnNo = turnNo;
    this.ply = ply;
  }

  static newGame(enabledTypes: ReadonlySet<PieceType> = BASE_PIECE_TYPES): GameState {
    const piecesList = buildPieces(enabledTypes);
    const pieces = new Map(piecesList.map((p) => [p.id, p]));
    const hand: [number[], number[]] = [[], []];
    for (const p of piecesList) hand[p.owner].push(p.id);
    const position = new Map<number, Pos | null>(piecesList.map((p) => [p.id, null]));
    return new GameState(pieces, new Map(), position, hand, [false, false], 0, 1, 0);
  }

  opponent(): Owner {
    return this.currentPlayer === 0 ? 1 : 0;
  }

  occupied(): ReadonlySet<string> {
    return new Set(this.board.keys());
  }

  stackHeightAt(pos: Pos): number {
    return this.board.get(posKey(pos))?.length ?? 0;
  }

  topPieceAt(pos: Pos): number | null {
    const stack = this.board.get(posKey(pos));
    return stack && stack.length > 0 ? stack[stack.length - 1] : null;
  }

  /** 0-indexed height of `pieceId` within its stack (0 = ground). */
  stackIndexOf(pieceId: number): number {
    const pos = this.position.get(pieceId);
    if (pos == null) throw new Error(`piece ${pieceId} is not on the board`);
    const stack = this.board.get(posKey(pos));
    if (!stack) throw new Error(`no stack at ${posKey(pos)}`);
    return stack.indexOf(pieceId);
  }

  /** Positions where `owner` currently has the top piece of the stack.
   * Deliberately *not* "every position any piece of theirs has ever
   * occupied" -- a piece buried under an opponent's beetle/mosquito no
   * longer controls that hex for placement-adjacency purposes (the
   * "don't place touching an opponent" rule), only whichever piece is
   * currently on top does. Confirmed against the Python engine's oracle:
   * using every piece's own position here (buried or not) wrongly
   * forbids/allows placements near a mixed-color stack. */
  boardPositions(owner: Owner): Set<string> {
    const positions = new Set<string>();
    for (const [key, stack] of this.board) {
      const topId = stack[stack.length - 1];
      if (this.pieces.get(topId)!.owner === owner) positions.add(key);
    }
    return positions;
  }

  piecesOnBoard(owner: Owner): number[] {
    const result: number[] = [];
    for (const [pid, piece] of this.pieces) {
      if (piece.owner === owner && this.position.get(pid) != null) result.push(pid);
    }
    return result;
  }

  /** Pillbug freeze rule: a piece that moved "last turn" can't be moved or
   * thrown this ply. The `ply - 1` comparison (rather than the arguably
   * more obvious `ply`) is copied verbatim from the Python engine's
   * `is_frozen`, which itself copies the reference Django app's exact
   * comparison for oracle-parity reasons -- see engine/state.py. */
  isFrozen(pieceId: number): boolean {
    return this.lastMovedPieceId === pieceId && this.lastMovedPly === this.ply - 1;
  }
}

/** A snapshot of "the position" for repetition detection: board contents
 * (including full stacks, since a beetle's height matters) plus each
 * hand's contents plus whose turn it is. Doesn't account for the pillbug
 * freeze state (which piece moved last) -- a pragmatic simplification,
 * since this only needs to bound self-play game length, not be
 * tournament-rules-perfect. Python's version returns a hashable tuple;
 * this returns the equivalent as a single string, since that's what's
 * usable as a JS Map key. */
export function positionKey(state: GameState): string {
  const boardEntries = Array.from(state.board.entries())
    .map(([key, stack]) => `${key}|${stack.join(",")}`)
    .sort();
  const hand0 = [...state.hand[0]].sort((a, b) => a - b).join(",");
  const hand1 = [...state.hand[1]].sort((a, b) => a - b).join(",");
  return `${boardEntries.join(";")}::${hand0}::${hand1}::${state.currentPlayer}`;
}
